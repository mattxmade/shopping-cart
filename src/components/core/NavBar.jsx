import React from "react";
import { NavLink } from "react-router-dom";

import PropTypes from "prop-types";
import errorLog from "../../logs/errorLog";

import "./NavBar.style.scss";

const validPaths = (paths) => {
  if (!paths.length) errorLog("NavBar: array is empty");

  paths.map((path) => {
    if (typeof path !== "string") errorLog("NavBar: path must be a string");
  });

  return true;
};

const capitaliseString = (string, start, end) =>
  string.slice(start, end).toUpperCase() + string.slice(end);

const processPath = (string) => {
  const pathToProcess = string;

  return pathToProcess.slice(0, 1) === "/"
    ? capitaliseString(pathToProcess, 1, 2)
    : capitaliseString(pathToProcess, 0, 1);
};

const NavBar = ({ paths, openNavBar }) => {
  if (!validPaths(paths)) return;

  // <li onClick={() => openNavBar(false)}/>

  return (
    <nav>
      <ul id="primary-navigation">
        {paths.map((path, index) => {
          return (
            <li key={index}>
              <NavLink className="nav-link--select" to={path}>
                {path === "/" ? "Home" : processPath(path)}
              </NavLink>
            </li>
          );
        })}
      </ul>
    </nav>
  );
};

export default NavBar;

NavBar.propTypes = {
  paths: PropTypes.array,
};

// .at() string method requires core-js | https://stackoverflow.com/a/59285424
// required for jest testing
