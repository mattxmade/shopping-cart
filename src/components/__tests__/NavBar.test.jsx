import React, { Fragment } from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { BrowserRouter } from "react-router-dom";

import "@testing-library/jest-dom";
import "@testing-library/dom";

import NavBar from "../core/NavBar";

const paths = ["/", "shop", "basket"];

describe("Nav component test suite", () => {
  it("Matches snapshot", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <NavBar paths={paths} />
      </BrowserRouter>
    );
    expect(asFragment).toMatchSnapshot();
  });

  it("Renders correct number of listitems", () => {
    render(
      <BrowserRouter>
        <NavBar paths={paths} />
      </BrowserRouter>
    );

    const navItems = screen.getAllByRole("listitem");
    expect(navItems.length).toEqual(paths.length);
  });

  it("Renders NavLink components with expected paths", async () => {
    render(
      <BrowserRouter>
        <NavBar paths={paths} />
      </BrowserRouter>
    );

    const navLinks = await screen.findAllByRole("link");
    expect(navLinks.length).toEqual(paths.length);

    await navLinks.map((link, index) => {
      const currentPath = paths[index];
      const href = link.href;

      const expectedResult = href.slice(-href.length - currentPath.length);

      expect(link.href).toBe(expectedResult);
    });
  });
});
