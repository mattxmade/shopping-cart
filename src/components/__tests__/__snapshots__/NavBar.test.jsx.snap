// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav component test suite Matches snapshot 1`] = `[Function]`;
